src/sensor_hx71x.c

```diff
@@ -0,0 +1,267 @@
+        hx71x_pulse_clocks(hx71x);
+        // read 2's compliment int bits
+        counts = (counts << 1) | gpio_in_read(hx71x->dout);
+    }
+
+    // bit bang 1 to 4 more bits to configure gain & channel for the next sample
+    for (uint8_t gain_idx = 0; gain_idx < hx71x->gain_channel; gain_idx++) {
+        hx71x_pulse_clocks(hx71x);
+    }
+
+    if (hx71x_is_data_ready(hx71x)) {
+        shutdown("HX71x Sensor reporting data ready after read");
+    }
+
+    hx71x_time_t time_diff = timer_read_time() - start_time;
+    if (time_diff >= MAX_READ_TIME) {
+        shutdown("HX71x Read took too long");
+    }
+
+    // extend 2's complement 24 bits to 32bits
+    if (counts >= 0x800000) {
+        counts |= 0xFF000000;
+    }
+
+    if (counts < -0x7FFFFF || counts > 0x7FFFFF) {
+        shutdown("HX71x Counts value out of range");
+    }
+    add_sample(hx71x, counts);
+
+    flush_samples(hx71x, oid);
+    hx71x_reschedule_timer(hx71x);
+}
+
+// Create a hx71x sensor
+void
+command_config_hx71x(uint32_t *args)
+{
+    struct hx71x_adc *hx71x = oid_alloc(args[0]
+                , command_config_hx71x, sizeof(*hx71x));
+    hx71x->timer.func = hx71x_event;
+    hx71x->pending_flag = 0;
+    uint8_t gain_channel = args[1];
+    if (gain_channel < 1 || gain_channel > 4) {
+        shutdown("HX71x gain/channel out of range 1-4");
+    }
+    hx71x->gain_channel = gain_channel;
+    hx71x->dout = gpio_in_setup(args[2], -1);
+    hx71x->sclk = gpio_out_setup(args[3], 0);
+}
+DECL_COMMAND(command_config_hx71x, "config_hx71x oid=%c gain_channel=%c"
+    " dout_pin=%u sclk_pin=%u");
+
+// start/stop capturing ADC data
+void
+command_query_hx71x(uint32_t *args)
+{
+    uint8_t oid = args[0];
+    struct hx71x_adc *hx71x = oid_lookup(oid, command_config_hx71x);
+    sched_del_timer(&hx71x->timer);
+    hx71x->pending_flag = 0;
+    hx71x->rest_ticks = args[1];
+    if (!hx71x->rest_ticks) {
+        // End measurements
+        return;
+    }
+    // Start new measurements
+    sensor_bulk_reset(&hx71x->sb);
+    hx71x_reschedule_timer(hx71x);
+}
+DECL_COMMAND(command_query_hx71x,
+             "query_hx71x oid=%c rest_ticks=%u");
+
+void
+command_query_hx71x_status(const uint32_t *args)
+{
+    uint8_t oid = args[0];
+    struct hx71x_adc *hx71x = oid_lookup(oid, command_config_hx71x);
+    const uint32_t start_t = timer_read_time();
+    uint8_t pending_bytes = 0;
+    pending_bytes = hx71x_is_data_ready(hx71x) * BYTES_PER_SAMPLE;
+    const uint32_t end_t = timer_read_time();
+    sensor_bulk_status(&hx71x->sb, oid, start_t, (end_t - start_t)
+                      , pending_bytes);
+}
+DECL_COMMAND(command_query_hx71x_status, "query_hx71x_status oid=%c");
+
+// Background task that performs measurements
+void
+hx71x_capture_task(void)
+{
+    if (!sched_check_wake(&wake_hx71x))
+        return;
+    uint8_t oid;
+    struct hx71x_adc *hx71x;
+    foreach_oid(oid, hx71x, command_config_hx71x) {
+        if (hx71x->pending_flag) {
```

Comment:
Same comment as before - I'm not sure what is intended with the pending_flag, as it is only ever set during sampling.  Something doesn't look correct.

Reply:
I think this is fixed now so it works like ADXL345